File: 2022-04/javascript/733_Flood_Fill.js
/**
 * @param {number[][]} image
 * @param {number} sr
 * @param {number} sc
 * @param {number} newColor
 * @return {number[][]}
 */
function floodFill(image, sr, sc, newColor) {
    // return floodFillBFS(image, sr, sc, newColor)
    return floodFillDFS(image, sr, sc, newColor)
}


function floodFillBFS(image, sr, sc, newColor) {
    const oldColor = image[sr][sc]
    if (oldColor === newColor) {
        return image
    }
    const r = image.length
    const c = image[0].length
    const stack = [{x: sr, y: sc}]
    while (stack.length > 0) {
        const {x, y} = stack.pop()
        if (image[x][y] == oldColor) {
            image[x][y] = newColor
            const dirs = [
                {x, y: y - 1},
                {x, y: y + 1},
                {x: x - 1, y},
                {x: x + 1, y},
            ]
            for (let {x, y} of dirs) {
                if (x >= 0 && x < r && y >= 0 && y < c) {
                    stack.push({x, y})
                }
            }
        }
    }
    return image
}


function floodFillDFS(image, sr, sc, newColor) {
    const oldColor = image[sr][sc]
    if (oldColor === newColor) {
        return image
    }
    const r = image.length
    const c = image[0].length

    function dfs(x, y) {
        if (x >= 0 && x < r && y >= 0 && y < c && image[x][y] === oldColor) {
            image[x][y] = newColor
            dfs(x, y - 1)
            dfs(x, y + 1)
            dfs(x - 1, y)
            dfs(x + 1, y)
        }
    }
    dfs(sr, sc)
    return image
}

File: 2022-04/python/733_Flood_Fill.py
class Solution:
    def floodFill(self, image: List[List[int]], sr: int, sc: int, new_color: int) -> List[List[int]]:
        return self.floodFillDFS(image, sr, sc, new_color)
        # return self.floodFillBFS(image, sr, sc, new_color)

    def floodFillDFS(self, image: List[List[int]], sr: int, sc: int, new_color: int) -> List[List[int]]:
        old_color, r, c = image[sr][sc], len(image), len(image[0])
        if old_color == new_color:
            return image
        def dfs(x: int, y: int):
            if 0 <= x < r and 0 <= y < c and image[x][y] == old_color:
                image[x][y] = new_color
                dfs(x - 1, y)
                dfs(x + 1, y)
                dfs(x, y - 1)
                dfs(x, y + 1)
        dfs(sr, sc)
        return image

    def floodFillBFS(self, image: List[List[int]], sr: int, sc: int, new_color: int) -> List[List[int]]:
        old_color, r, c = image[sr][sc], len(image), len(image[0])
        if old_color == new_color:
            return image
        stack = [(sr, sc)]
        while stack:
            x, y = stack.pop()
            if image[x][y] == old_color:
                image[x][y] = new_color
                dirs = [
                    (x - 1, y),
                    (x + 1, y),
                    (x, y - 1),
                    (x, y + 1),
                ]
                for x, y in dirs:
                    if 0 <= x < r and 0 <= y < c:
                        stack.append((x, y))
        return image